devito/finite_differences/derivative.py

```diff
@@ -0,0 +1,192 @@
+    >>> grid = Grid((10, 10))
+    >>> x, y = grid.dimensions
+    >>> u = Function(name="u", grid=grid, space_order=2)
+    >>> Derivative(u, x)
+    Derivative(u(x, y), x)
+
+    This can also be obtained via the differential shortcut
+
+    >>> u.dx
+    Derivative(u(x, y), x)
+
+    You can also specify the order as a keyword argument
+
+    >>> Derivative(u, x, deriv_order=2)
+    Derivative(u(x, y), (x, 2))
+
+    Or as a tuple
+
+    >>> Derivative(u, (x, 2))
+    Derivative(u(x, y), (x, 2))
+
+    Once again, this can be obtained via shortcut notation
+
+    >>> u.dx2
+    Derivative(u(x, y), (x, 2))
+    """
+
+    _state = ('expr', 'dims', 'side', 'stagger', 'fd_order', 'transpose')
+
+    def __new__(cls, expr, *dims, **kwargs):
+        if type(expr) == sympy.Derivative:
+            raise ValueError("Cannot nest sympy.Derivative with devito.Derivative")
+        if not isinstance(expr, Differentiable):
+            print(expr, type(expr), dims, kwargs)
+            raise ValueError("`expr` must be a Differentiable object")
+
+        # Check `dims`. It can be a single Dimension, an iterable of Dimensions, or even
+        # an iterable of 2-tuple (Dimension, deriv_order)
+        if len(dims) == 0:
+            raise ValueError("Expected Dimension w.r.t. which to differentiate")
+        elif len(dims) == 1:
+            if isinstance(dims[0], Iterable):
+                # Iterable of Dimensions
+                if len(dims[0]) != 2:
+                    raise ValueError("Expected `(dim, deriv_order)`, got %s" % dims[0])
+                orders = kwargs.get('deriv_order', dims[0][1])
+                if dims[0][1] != orders:
+                    raise ValueError("Two different values of `deriv_order`")
+                new_dims = tuple([dims[0][0]]*dims[0][1])
+            else:
+                # Single Dimension
+                orders = kwargs.get('deriv_order', 1)
+                if isinstance(orders, Iterable):
+                    orders = orders[0]
+                new_dims = tuple([dims[0]]*orders)
+        else:
+            # Iterable of 2-tuple, e.g. ((x, 2), (y, 3))
+            new_dims = []
+            orders = []
+            d_ord = kwargs.get('deriv_order', tuple([1]*len(dims)))
+            for d, o in zip(dims, d_ord):
+                if isinstance(d, Iterable):
+                    new_dims.extend([d[0] for _ in range(d[1])])
+                    orders.append(d[1])
+                else:
+                    new_dims.extend([d for _ in range(o)])
+                    orders.append(o)
+            new_dims = as_tuple(new_dims)
+            orders = as_tuple(orders)
+
+        # Finite difference orders
+        fd_orders = kwargs.get('fd_order', tuple([expr.space_order]*len(dims)))
+        if len(dims) == 1 and isinstance(fd_orders, Iterable):
+            fd_orders = fd_orders[0]
+        # SymPy expects the list of variable w.r.t. which we differentiate to be a list
```

Comment:
a blank line between this comment and the block above wouldn't harm

Reply:
DOne